### ISSUE TITLE
Datasources and persistence units with multiple "alternatives" listed at build time and selected at runtime

### ISSUE THREAD

**Issue by yrodiere** [quarkusio/quarkus#43656]
## Problem statement

Some applications are built once, then distributed as a single binary (container image) and consumed/deployed by third-parties (users, customers) in various environments.

Such binaries need to work in multiple environments, and that means in particular that datasources and persistence units may need to work with different database kinds (e.g. Oracle and Postgres) depending on the deployment.

Currently, application developers have two solutions to address this:

1. Declare multiple named persistence units and datasources, one per db-kind, make them [active/inactive](https://quarkus.io/guides/hibernate-orm#persistence-unit-active) based on configuration profiles, have a custom PU/datasource CDI bean whose producer would have point to the active PU/datasource at runtime (see previous link).
2. Produce a "mutable jar" and have users/customers execute [re-augmentation](https://quarkus.io/guides/reaugmentation#2-re-augment-your-application-with-a-different-build-time-configuration).

This has several downsides, however:

* Either solution can be quite complex to implement for application developers.
* Solution 1 implies an distributed container image that contains *all* the alternatives -- but that's the bare minimum to address such use case.
* Solution 1 doesn't work with Panache -- which [requires each entity to be bound to at most one persistence unit](https://quarkus.io/guides/hibernate-orm-panache#multiple-persistence-units).
* Solution 2 implies a distributed container image that is much larger (includes deployment JARs and who knows what else needed for reaugmentation).
* Solution 2 implies a intermediary "rebuild" step that will either slow down startup and make the container image much larger (if executed on container startup) or make consumption of the container image much more complex (if executed manually by the users/customers to derive a final container image).
* Solution 2 is not compatible with native builds.

## Proposal

We could expose more guided, better integrated ways of developing and deploying of such applications.

----

Related discussion: https://github.com/quarkusio/quarkus/discussions/42391

## If we prefer solution 1 above (active/inactive)

We could provide a built-in concept of "alternatives" in the configuration of persistence units or datasources:

```properties
# Defining the default datasource with multiple alternatives.
# In practice the extension will generate one set of bytecode/bean per alternative,
# and use only one at runtime.
quarkus.datasource.jdbc.min-size=10
quarkus.datasource.alternatives."db2".db-kind=db2
quarkus.datasource.alternatives."db2".jdbc.max-size=60
quarkus.datasource.alternatives."postgresql".db-kind=postgresql
quarkus.datasource.alternatives."postgresql".jdbc.max-size=40

# Runtime profiles allow selecting an alternative
%db2.quarkus.datasource.selected-alternative=db2
%pg.quarkus.datasource.selected-alternative=pg

# Hibernate ORM consumes the default datasource,
# so it will automatically generate one alternative per datasource,
# and automatically select the correct one depending on the selected datasource at runtime
quarkus.hibernate-orm.database.generation=create
```

Main downsides:

1. This is very ad-hoc and requires per-extension implementation. We may provide some basic scaffoding in Quarkus (e.g. standard build items and helpers around configuration), but most of the heavy lifting will have to be implemented per extension.
2. \<insert your concerns here>

## If we prefer solution 2 above (re-augmentation)

We could provide a *simple* process and technical solution that could be enabled *simply* (e.g. configuration flag) by application developers, and could by deployed *simply* by users/customers, triggering re-augmentation *only once*:

* We could have Quarkus generate a multi-stage docker file for users/consumers to consume. Perhaps it is simple enough for Openshift (which I think allows defining the application to deploy using a Dockerfile), but it probably isn't for everyday docker/podman/testcontainers usage.
* We could have Quarkus generate an operator as part of the build, whose only task would be to trigger the build and. Not sure if it makes sense or if it's even feasible, though.
* Open to other ideas... ?

Main downsides:

1. This is still quite complex.
2. This still implies a slow first deployment.
3. This still requires large distributed applications (they include deployment JARs).
4. This still wouldn't support native builds.

----

cc @maxandersen @cescoffier @geoand @gsmet and anyone you think would be interested: your ideas could help here...

**Comment by FroMage** [comment]
> Solution 1 doesn't work with Panache 

See for a possible solution for Panache 1 and repositories https://github.com/quarkusio/quarkus/issues/43341

Panache 2 should have something better, to also make this possible.

**Comment by shepherd85** [comment]
I _really_ need something along the lines of Solution 1 + #43341.  Reaugmentation is not preferable, as I am using Quinoa to bundle the frontend.

**Comment by yrodiere** [comment]
> I _really_ need something along the lines of Solution 1 + [#43341](https://github.com/quarkusio/quarkus/issues/43341). Reaugmentation is not preferable, as I am using Quinoa to bundle the frontend.

Could you clarify why you _need_ it?

I can certainly understand _wanting_ it, because the current situation is not exactly ideal, but none of the suggestions here provide anything new: you can currently do it by hand. It's inconvenient, but certainly possible (people do it already).

At least that's what I thought, so I'd love to hear of how this could help address new usecases.

**Comment by yrodiere** [comment]
That's a valid way to expose the feature, though it has downsides (what if two datasources are active?) and the problem is unfortunately not limited to how we present things (see mention of Hibernate ORM in the description).

More to the point:

> Could you clarify why you _need_ it?
> 
> I can certainly understand _wanting_ it, because the current situation is not exactly ideal, but none of the suggestions here provide anything new: you can currently do it by hand. It's inconvenient, but certainly possible (people do it already).
> 
> At least that's what I thought, so I'd love to hear of how this could help address new usecases.

**Comment by shepherd85** [comment]
Sorry - didn't see your question until after posting.  This may be a bit long, and perhaps better suited for the discussion thread I found referencing this issue.  Also, forgive my ignorance on how things work under the hood.

> Could you clarify why you need it? ... you can currently do it by hand.

Honestly, a bit petulant on my behalf, but I was a bit frustrated having only recently learned that this very common use case is not directly supported, and the setback of having to pull Panache out, but supporting multiple datasources with a simple configuration change is a hard requirement, and reaugmentation is not great for reasons below.

> I'd love to hear of how this could help address new usecases.

I am building an API service targeting hosted and local installs.  For hosted installs, it is bundled with a UI using Quinoa.  Reagumentation is not a great option here because it requires mutable jars, causing the entire UI source to be included and rebuilt from source.  It also precludes using native images.

For local installs, it is intended to service an Electron app, so it is not bundled with the UI.  The application needs to either support a local database, or connect to a central database.  Reaugmentation would work better here, but still not great as now I have to install a JVM alongside the electron app. 

So, can it be done other ways?  Sure.  I'm currently rewriting all my entities and APIs to use the "active" datasource examples using profiles.  Still, it shouldn't be this hard, and it just seems odd that this very common scenario (runtime configuration of database backends) was not given more thought.

So now, back to this:
> That's a valid way to expose the feature, though it has downsides (what if two datasources are active?) 

Perhaps "proxy" is the wrong `db-kind`.  Perhaps `delegating`?  Ex:
```
quarkus.datasource.db-kind=delegating
# first "active" datasource is the chosen delegate
quarkus.datasource.delegates=<list of named datasources: app-mssql,app-postgres>

quarkus.datasource."app-mssql".<...>=...
quarkus.hibernate-orm."app-mssql".<...>=...
```

Whether proxy or delegating, the idea is that it only ever delegates to a single named datasource.  That's why I had the `active` config on the proxy datasource.  But perhaps it just needs to delegate to the first "active" datasource configured as one of it's sources.

The individual sources can all be active, or not.  They can be used individually.  It's just that the proxy datasource delegates a single "active" datasource, as configured.

> and the problem is unfortunately not limited to how we present things (see mention of Hibernate ORM in the description).

Perhaps I'm too naive, but my thinking also involves implementing a delegating PU.  Runtime should be transparent, build time would require treating entities bound to the delegating PU as being bound to each of the sources, and whatever necessary configuration steps are needed to build the various PU alternatives.  To Panache at runtime, however, it would still consider the entities as configured to a single PU, so that _just works_, as they are all bound to the delegate.

**Comment by yrodiere** [comment]
> the setback of having to pull Panache out

FWIW there are plans to support multiple persistence units in Panache. Possibly in the current implementation but for the Repository approach only, and definitely in the longer term in Panache 2.
See https://github.com/quarkusio/quarkus/issues/43341

> The individual sources can all be active, or not. They can be used individually. It's just that the proxy datasource delegates a single "active" datasource, as configured.

Makes sense, this simplifies things compared to my initial proposal since the "delegating" datasource is not a template.

Though not being a template comes with downsides: we'd need to check (somehow) that nobody tries to set any property (jdbc url, pool size, ...) on a "delegating" datasource, because that would not work. I'm not sure that's doable easily and reliably (not through a hack) in Smallrye Config.

Regarding the selection of the delegate, being explicit about it seems better to me, as it removes the need for error cases if two different "delegates" are active.

Pretty sure that's not the end of our trouble, but we can certainly work on problems as they arise :)

BTW the `db-kind` seems superfluous, we could just have this:

```
# Build-time
quarkus.datasource.possible-delegates=h2,mssql,postgres

# Runtime
%dev.quarkus.datasource.selected-delegate=h2
%pg.quarkus.datasource.selected-delegate=postgres
# If selected-delegate is not set, the delegating datasource is inactive
```

(`possible-delegates`/`selected-delegate` being placeholder names intended to go through bikeshedding later)

> Perhaps I'm too naive, but my thinking also involves implementing a delegating PU. Runtime should be transparent, build time would require treating entities bound to the delegating PU as being bound to each of the sources, and whatever necessary configuration steps are needed to build the various PU alternatives. To Panache at runtime, however, it would still consider the entities as bound to a single PU, so that just works, as they are all bound to the delegate.

That's the gist of it, yes.

Though whatever we do, we'll need to look through the code for (numerous) places that pull information from the (build-time) list of datasources: Hibernate ORM, but also Flyway, Liquibase, and perhaps others.

Also there might be more challenges for reactive datasources, where I don't think we have a centralized place to implement this delegation stuff, and the exposed type of the datasource differs depending on the db-kind.

--- 

Anyway, like I was trying to say: this is doable, but not simple as there's quite some work -- and maybe refactoring -- involved, and we'll need that work to be done rigorously (to make sure everything dealing with datasources still work). That's probably a few of the reasons nobody tried to address this before.

If someone is interested in tackling this sooner rather than later, I can help them with problems along the way, and I can help with reviews. The first step, considering this is a problem that may affect other areas of Quarkus, is probably to initiate a discussion on  quarkus-dev@googlegroups.com to make sure everyone agrees with the approach.

**Comment by shepherd85** [comment]
> BTW the db-kind seems superfluous
If implemented in the core "datasources" 
Is there anything that comes to mind that would make it technically infeasible to begin implementing this as a 3rd party JDBC extension providing the "delegating" db-kind?  That seems to me to be the most intuitive path ATM.

> we'd need to check (somehow) that nobody tries to set any property (jdbc url, pool size, ...) on a "delegating" datasource

The driver extension could issue warnings/throw exceptions if configured with any of the **.jdbc** properties.

If you think it is feasible, I'm willing to give it a try - at least for non-reactive datasources to begin with.  I will also reach out to the dev group.

**Comment by yrodiere** [comment]
> > BTW the db-kind seems superfluous
> 
> Is there anything that comes to mind that would make it technically infeasible to begin implementing this as a 3rd party JDBC extension providing the "delegating" db-kind? That seems to me to be the most intuitive path ATM.

Yes there is: integration with Hibernate ORM, which I mentioned above. This extension needs to be aware of the db-kind (among others) of the datasource used by a given persistence unit -- at build time. We're talking about changes to (core) `BuildItems`.

> > we'd need to check (somehow) that nobody tries to set any property (jdbc url, pool size, ...) on a "delegating" datasource
> 
> The driver extension could issue warnings/throw exceptions if configured with any of the **.jdbc** properties.

Yes, that's what I was suggesting -- though I doubt that this is limited to `.jdbc` properties, since `.username` for example is in the same case. The question is how to do that reliably, and in a future-proof way (if more properties are added).

> If you think it is feasible

It is, though it'll likely be an interesting challenge :)

> I'm willing to give it a try - at least for non-reactive datasources to begin with. I will also reach out to the dev group.

Thanks! Looking forward to that.

**Comment by yrodiere** [comment]
> > If you think it is feasible
> 
> It is, though it'll likely be an interesting challenge :)

To be clear I meant the feature can be implemented in the core. I don't think it's feasible as a third-party extension -- at least not without breaking integration with Hibernate ORM and perhaps others (not sure about Flyway/Liquibase).

**Comment by FroMage** [comment]
@yrodiere is there any reason we can't use multi-tenancy for that? One tenant per supported DB kind, and the tenant resolver would return the selected tenant ID at runtime, so in practice you only ever use one tenant at runtime, but each runtime can start with a different db-kind/tenant.

Unless multi-tenancy is limited to the same db kind?

This would have the benefit of already working OOTB and supported by Panache.

**Comment by yrodiere** [comment]
> Unless multi-tenancy is limited to the same db kind?

Yep. Very core design principle in Hibernate ORM itself. One PU = one dialect. Every single aspect of the "internal metamodel" is affected by this design principle. E.g. each "attribute" is mapped to a single JDBC type, whereas supporting multiple DBs would involve some "alternative" handling instead, with a different JDBC type depending on the DB.

> This would have the benefit of already working OOTB and supported by Panache.

It would, if it was possible -- unfortunately it's not :) Panache 2 + multiple PUs seems more realistic to me.

**Comment by FroMage** [comment]
Pity that tenants and datasources are so similar and yet separate :-/

**Comment by yrodiere** [comment]
> Pity that tenants and datasources are so similar and yet separate :-/

Agreed in general, but I don't think that matters here: even if multitenancy was handled at the datasource level, we'd still have a problem in Hibernate ORM which needs to know, at build time, the dialect to use to talk to a datasource. If that datasource can switch from PostgreSQL to Oracle at runtime, Hibernate ORM needs to be aware of that and adapt. Much wider implications than multitenancy.

**Comment by yrodiere** [comment]
FTR this was discussed to some extent on the mailing list: https://groups.google.com/g/quarkus-dev/c/GtQwS7O0KA0/m/zes40eS2EAAJ

We also happened to discuss it during a team meeting. I suggested some core concept of “build variants” in a single binary, i.e. handling all the complexity of "alternatives" at the Quarkus "build chain" level, but found no buyer, so that won't happen :)

So it looks like we're going to go with specific code to handle this for datasources only, and more specific code to handle the consequences on consuming extensions (Hibernate ORM/Reactive for sure, Flyway/Liquibase and others maybe). It should look something like this in the configuration:

```properties
# Build-time
quarkus.datasource.alternatives."pg".db-kind=postgresql
quarkus.datasource.alternatives."oracle".db-kind=oracle

# Runtime (e.g. commandline argument)
%pg.quarkus.datasource.selected-alternative=pg
%oracle.quarkus.datasource.selected-alternative=oracle
```

The extent of what can be configured per alternative, and what cannot, is still TBD. I would suggest that we simply allow everything, and use "non-alternative" config as a "default" to be applied to all alternatives. E.g.:

```
quarkus.datasource.alternatives."pg".db-kind=postgresql
quarkus.datasource.alternatives."oracle".db-kind=oracle
# Applies to both pg and oracle
quarkus.datasource.max-size=14
```

That being said, this may still require twisting the "build step" system -- which @dmlloyd will be interested in -- and we'll likely need some help on exposing "alternatives" of beans in Arc -- which @Ladicek offered to provide. So we'll probably have some meeting after someone started looking a bit more into what the challenges are exactly.
Clément@cescoffier ~~is forcing us to~~ kindly suggested that we create a [working group](https://quarkus.io/working-groups/) for that, to give more visibility on progress/challenges. @cescoffier I didn't find anything about the process to start new working groups on the [landing page](https://quarkus.io/working-groups/), can you please offer a hint?
